### PR TITLE
Prevent IO operations from being called over and over

### PR DESCRIPTION
**Pull request by Thodor12**
Regarding a problem that was mentioned a few days ago in the Discord, we found out that in the build tool (possibly more places) there are continuous image updates through some form of onUpdate, or scrolling list update.

These updates run every tick and cause IO operations to be fired off down the line here, whilst not required at all.

Probable fix for this is to simply check the input arguments into `setImage` and block the image loading again if the arguments remain the same.

### CODE REVIEW

**Comment by Nightenom** [comment]
well, the issue is known for ages, you should not update unless necessary in scrolling list data provider (Ray was told + knew + it's Ray xd), the problem is current impl of whole scroll group, rework is planned
not sure if we want to check for every possible setter if the value has changed or not (assuming the setter causes some non-trivial logic to run)

**Comment by Raycoms** [comment]
> well, the issue is known for ages, you should not update unless necessary in scrolling list data provider (Ray was told + knew + it's Ray xd), the problem is current impl of whole scroll group, rework is planned not sure if we want to check for every possible setter if the value has changed or not (assuming the setter causes some non-trivial logic to run)

 Its mostly image loading, so I think this is def a good workaround for now.

**Comment by Nightenom** [comment]
its also text rendering for sure, not sure if any other setters
what i would rather see as simple yet more powerful fix - detach scrolling list data provider from update and make it callable method
it kinda easy, just delete this [line in ScrollingList](https://github.com/ldtteam/BlockUI/blob/version/latest/src/main/java/com/ldtteam/blockui/views/ScrollingList.java#L88)
yeah it's breaking change unlike setter guard, and will require mcol changes

**Comment by Raycoms** [comment]
Text rendering doesn't do io from disk. But I think whatever we do, we want both, better list handling and proper double checks. But breaking changes we should reserve to 1.19.4

Sent from Proton Mail mobile

-------- Original Message --------

**Review by Nightenom — CHANGES_REQUESTED** [review]
definitely change both things to sth like if (resLoc changed) loadImageDimensions(resLoc);
other arguments are irrelevant for guarding the IO

**Comment by Thodor12** [comment]
> definitely change both things to sth like if (resLoc changed) loadImageDimensions(resLoc); other arguments are irrelevant for guarding the IO

The imageHeight and imageWidth do not affect the image as it's being loaded?
Is that just used in a transformation post image load?

**Comment by Nightenom** [comment]
> The imageHeight and imageWidth do not affect the image as it's being loaded? Is that just used in a transformation post image load?

variable naming is a bit problematic in there because there many widths and many offsets, this should be used as real blit area

**Comment by Thodor12** [comment]
Updated the logic @Nightenom if it's good this way I'll update the 1.18.2 PR too

File: src/main/java/com/ldtteam/blockui/controls/ButtonImage.java [hunk 1]
```diff
@@ -198,13 +198,16 @@ private void loadImageDisabledDimensions(final ResourceLocation rl)
...
     public void setImage(final ResourceLocation loc, final int offsetX, final int offsetY, final int w, final int h)
     {
+        if (!Objects.equals(loc, image))
+        {
+            loadImageDimensions(loc);
+        }
+
         image = loc;
         imageOffsetX = offsetX;
         imageOffsetY = offsetY;
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/blockui/controls/ButtonImage.java:208]: should that if be after the value setting like how it was before?

Comment (Nightenom) [reply src/main/java/com/ldtteam/blockui/controls/ButtonImage.java:208]: doesnt matter

**Comment by Nightenom** [comment]
please also apply to Image class (all call sites of getImageDimensions or whats the method name)

**Comment by Thodor12** [comment]
Also added it to highlight and disabled images for the buttons, seeing as those had the same method structure

**Review by Nightenom — APPROVED** [review]

**Pull request merged**